packages/p2p-media-loader-hlsjs/src/services/segment-mananger.ts

```diff
@@ -0,0 +1,138 @@
+import {
+  Parser,
+  type PlaylistManifest,
+  type MasterManifest,
+  type Segment as ParserSegment,
+} from "m3u8-parser";
+
+export class SegmentManager {
+  videoPlaylists: VideoPlaylistsContainer = new VideoPlaylistsContainer();
+  audioPlaylists: AudioPlaylistsContainer = new AudioPlaylistsContainer();
+
+  processPlaylist(content: string, requestUrl: string, responseUrl: string) {
+    const parser = new Parser();
+    parser.push(content);
+    parser.end();
+
+    const { manifest } = parser;
+    if (TypeGuard.isMasterManifest(manifest)) {
+      this.videoPlaylists.setFromMasterManifest(responseUrl, manifest);
+      this.audioPlaylists.setFromMasterManifest(responseUrl, manifest);
+    } else {
+      const playlist =
+        this.videoPlaylists?.getPlaylist(responseUrl) ??
+        this.audioPlaylists?.getPlaylist(responseUrl);
+      if (playlist) playlist.setSegments(manifest.segments);
+    }
+  }
+}
+
+abstract class PlaylistsContainer {
+  masterPlaylistUrl?: string;
+  playlists: Playlist[] = [];
+  playlistMap: Map<string, Playlist> = new Map<string, Playlist>();
+
+  abstract setFromMasterManifest(
+    masterManifestUrl: string,
+    masterManifest: MasterManifest
+  ): void;
+
+  getPlaylist(playlistUrl: string): Playlist | null {
+    return this.playlistMap.get(playlistUrl) ?? null;
+  }
+
+  getPlaylistBySegmentUrl(segmentUrl: string): Playlist | null {
+    for (const playlist of this.playlists) {
+      if (playlist.segmentsMap.has(segmentUrl)) return playlist;
+    }
+    return null;
+  }
```

Comment:
`undefined` is usually better than `null`. For example, the requirement for developers who create TypeScript language: https://github.com/Microsoft/TypeScript/wiki/Coding-guidelines#null-and-undefined